### ISSUE TITLE
Tensorboard logging in multi-gpu setting not working properly?

### ISSUE THREAD

**Issue by aitor-garcia-p** [williamFalcon/pytorch-lightning#230]
Hi there :)

I have a question (that may be an issue with the code or just my ignorance).
(b.t.w. I am using the latest version, pytorch-lightning==0.4.9)

If I set the trainer 

`trainer = Trainer(experiment=exp, gpus=[0])`

I can see the corresponding logging (scalars and hyperparameters) in Tensorboard.
If I change it to distributed training (keeping the rest of the code unchanged) :

`
trainer = Trainer(experiment=exp, gpus=[0,1], distributed_backend='ddp')`

The Tensorboard logging stops working at least for scalars and hyperparameters, I see nothing except the experiment name.

In both cases 'exp' is a Experiment instantiated like this:

`    exp = Experiment(save_dir=/SOME/PATH, name=NAME, version=VERSION, description=DESCRIPTION)`

This picture illustrates the problem.

![Untitled](https://user-images.githubusercontent.com/2588285/65029817-b175d500-d93e-11e9-9f3f-3e350d4188d4.png)

In the picture the red arrows point to the "distributed" experiment, with no drawing in the chart. The other two (the ones that appear in the chart) are the very same, except than the run in single GPU.

Am I missing something or do I need to add extra configuration to make the logging work in multi-gpu with the ddp setting? Or is it a bug?

Thank you! :)

**Comment by williamFalcon** [comment]
could you post the full code that uses the trainer?

**Comment by aitor-garcia-p** [comment]
Do you mean this part?

```
def main(hparams):
    model = MyCustomizedLightningModel(hparams)

    # harcoded path here
    exp = Experiment(save_dir='/DATA/exp3',
                     name='classif',
                     version='0.1',
                     description='description goes here')

    exp.argparse(hparams)

    trainer = Trainer(min_nb_epochs=1,
                      max_nb_epochs=5, nb_sanity_val_steps=1, experiment=exp, gpus=[0,1], distributed_backend='ddp')
    trainer.fit(model)
```

Only when I remove the "distributed_backend='ddp'" (and the second gpu from the list) to make it single-gpu it does fill the charts and hparams report in tensorboard, like in the picture above.

**Comment by williamFalcon** [comment]
not sure what it would be, but try an integer and make sure to change it between experiments

```
    exp = Experiment(save_dir='/DATA/exp3',
                     name='classif',
                     version=1,  # increase every new experiment
                     description='description goes here')
```

**Comment by aitor-garcia-p** [comment]
I have tried with integer version numbers, but no luck.

![Untitled](https://user-images.githubusercontent.com/2588285/65042639-8cdb2680-d959-11e9-901b-dd72c30426a6.png)

The tensorboard writing/no-writing seems to be triggered by the "distributed_backend='ddp'" absence/presence in the Trainer instantiation.
Maybe when it is in distributed mode it writes to some other default location? Or there is an if/else related to the 'ddp' that prevents the SumaryWriter add the values to the report?

**Comment by williamFalcon** [comment]
not sure... are you using jupyter notebook?

**Comment by aitor-garcia-p** [comment]
No, I am coding using an IDE.

Don't worry anyway, sure there are more urgent tasks. For me it's ok for now, I can do most of the trainings on a single GPU :-)
Just wanted to know if I was doing something wrong. Maybe it has to do with something in my system. If I discover anything suspicious I will inform here.
If more people starts reporting about this problem then you can invest more attention on this.
Thanks for your time!

**Comment by seungwooYoo** [comment]
Actually I also observed the similar things when training with 'ddp' option. Seems that tfevents file does not correctly written (no file size increased during the training). I also can do most of thing using 'dp' option now....

**Comment by seungwooYoo** [comment]
Seems like self.proc_rank is not well set in 'ddp'. I print self.proc_rank in "trainer" and found all process has proc_rank=0 which makes self.experiment to be working as debug mode.

**Comment by williamFalcon** [comment]
the code you show is from an old version of lightning. update to the latest to see if you still have an issue

**Comment by williamFalcon** [comment]
Ok, submitted a PR. 
Can you install this version and verify it works now?
```
git fetch origin
git checkout -b fix_tb_logger_rank origin/fix_tb_logger_rank
```

**Comment by seungwooYoo** [comment]
I tested with update PR and found it generates several meta files in start. But I found there are another crashes comes from RandomSampler set_epoch. I added the error mssg in the below. 

![image](https://user-images.githubusercontent.com/1920881/65205934-e746c500-dac8-11e9-8427-ba5ee23c79b4.png)

**Comment by seungwooYoo** [comment]
Hi thanks for updating. Frankly, I observed another crash which may not be related to above issues. I uploaded the error mssg. 

![image](https://user-images.githubusercontent.com/1920881/65207071-8e2c6080-dacb-11e9-8942-a5df5f1f575b.png)

**Comment by williamFalcon** [comment]
yeah, you are setting up distributed sampler wrong. look at the pytorch and lightning docs.

use distributedsampler and don’t use the shuffle flag

**Comment by seungwooYoo** [comment]
Thanks for quick answer. :) I found that updated PR works great 👍 
![image](https://user-images.githubusercontent.com/1920881/65207598-08a9b000-dacd-11e9-810d-0fd1c1e0b816.png)

**Comment by samhumeau** [comment]
An additional related remark, which I think fits the title of this issue.

Right now pytorch-lightning seems to create several tfevent files in the multi-gpu ddp way:
e.g. for 2 GPUs:
```
-rw-rw-r--. 1 sam sam   40 Sep 19 08:11 events.out.tfevents.1568880714.google2-compute82.3156.0
-rw-rw-r--. 1 sam sam 165K Sep 19 08:22 events.out.tfevents.1568880716.google2-compute82.3186.0
-rw-rw-r--. 1 sam sam   40 Sep 19 08:11 events.out.tfevents.1568880718.google2-compute82.3199.0
```
I suppose the first one is created by the main process and the next 2 are created by the 2 DDP processes (one per GPU). Unfortunately, the actual events are not logged in the last created one, and that confuses tensorboard, cf https://github.com/tensorflow/tensorboard/issues/1011 

I have to restart tensorboard if I want to see the new data.

**Comment by williamFalcon** [comment]
@samhumeau we can open a new issue for this. When you init the experiment it creates a file handle. The solution is to make sure to init the experiment from the master gpu only. However there needs to be a way for this to happen so the user doesn't think about it.
